src/music/Scales.ts

```diff
@@ -0,0 +1,140 @@
+import { mod } from "../util/Util.js";
+
+import { Arrays } from "../util/Arrays.js";
+
+export type Scale = number; // scales are encoded with numbers in Ian Ring's format
+
+export interface ScaleQuery {
+    imperfections?: [number, number]; // allowed range for number of imperfections
+    hemitones?: [number, number]; // allowed range for number of hemitones
+    chord?: Scale; // chord that the scale must contain
+    notes?: [number, number]; // allowed range for number of notes in the scale
+}
+
+export namespace Scales {
+
+    export function createScaleFromPitchClass(pitches: number[]): Scale {
+        if (!pitches.includes(0)) {
+            throw new Error("scales must have a root");
+        }
+        return pitches.map(pitch => 1 << mod(pitch, 12)) // mod by 12 to fit into octave
+            .filter((val, i, arr) => arr.indexOf(val) === i) // remove duplicates
+            .reduce((acc, curr) => acc | curr); // combine into final bit vector
```

Comment:
Give `reduce` a default accumulator `.reduce(..., 0)`, since `0` is the identity under the `|` monoid here.